File: midi-encoder.js
import { pathToFileURL } from "node:url";

import {
  buildEvent,
  serialiseEvent,
  CHANNEL,
  PITCH_DOMAIN,
  VELOCITY,
  DURATION,
  pitchToNoteName
} from "./midi-grammar.js";
import { mxitCompress } from "./mxit-compressor.js";

const ACTION_DOMAINS = new Set([
  PITCH_DOMAIN.REQUEST,
  PITCH_DOMAIN.COMMAND,
  PITCH_DOMAIN.BUILD,
  PITCH_DOMAIN.FIX,
  PITCH_DOMAIN.SEARCH,
  PITCH_DOMAIN.SEND,
  PITCH_DOMAIN.SAVE
]);

function normaliseRole(event) {
  if (event.role) {
    return String(event.role).toLowerCase();
  }

  const matchedEntry = Object.entries(CHANNEL).find(([, value]) => {
    return value === event.channel;
  });

  return matchedEntry ? matchedEntry[0].toLowerCase() : "user";
}

function describeMeaning(event, role) {
  const domain = event.domain ?? "UNKNOWN";

  if (role === "user" && ACTION_DOMAINS.has(event.pitch)) {
    return `${domain} request`;
  }

  if (role === "assistant") {
    return `${domain} response`;
  }

  if (event.pitch === PITCH_DOMAIN.QUESTION) {
    return "QUESTION prompt";
  }

  return `${domain} message`;
}

function describeIntensity(velocity) {
  if (velocity >= VELOCITY.URGENT) {
    return `URGENT (${velocity})`;
  }
  if (velocity >= VELOCITY.EMPHASIS) {
    return `EMPHASIS (${velocity})`;
  }
  if (velocity >= VELOCITY.NORMAL) {
    return `NORMAL (${velocity})`;
  }
  if (velocity >= VELOCITY.SOFT) {
    return `SOFT (${velocity})`;
  }
  if (velocity >= VELOCITY.WHISPER) {
    return `WHISPER (${velocity})`;
  }
  return `SILENT (${velocity})`;
}

function describeImportance(duration) {
  if (duration >= DURATION.ANCHOR) {
    return `${DURATION.ANCHOR}ms`;
  }
  return `${duration}ms`;
}

export function encode(rawText, role = "user") {
  const compressed = mxitCompress(rawText);
  return buildEvent(rawText, compressed.text, role);
}

export function decode(event) {
  const role = normaliseRole(event);

  return {
    role,
    meaning: describeMeaning(event, role),
    intensity: describeIntensity(event.velocity ?? VELOCITY.NORMAL),
    importance: describeImportance(event.duration ?? DURATION.NORMAL),
    text: event.compressed ?? mxitCompress(event.raw ?? "").text
  };
}

export function encodeLog(messages = []) {
  return messages.map(({ text, role = "user" }) => encode(text, role));
}

export function logToContext(events = []) {
  return events.map((event) => serialiseEvent(event)).join("\n");
}

export function toPianoRoll(events = [], width = 32) {
  if (events.length === 0) {
    return "(empty piano roll)";
  }

  const columnCount = Math.max(1, Number(width) || 32);
  const visibleEvents = events.slice(-columnCount);
  const pitches = [...new Set(visibleEvents.map((event) => event.pitch))].sort((a, b) => b - a);

  return pitches.map((pitch) => {
    const label = pitchToNoteName(pitch).padEnd(3, " ");
    const columns = visibleEvents
      .map((event) => (event.pitch === pitch ? "*" : " "))
      .join("")
      .padEnd(columnCount, " ");

    return `${label} | ${columns} |`;
  }).join("\n");
}

const isDirectRun = Boolean(process.argv[1]) &&
  pathToFileURL(process.argv[1]).href === import.meta.url;

if (isDirectRun) {
  const event = encode(
    "I am going to build the system tomorrow because I need to know how it works",
    "user"
  );

  console.assert(event.role === "user", "Role should be preserved");
  console.assert(event.domain === "BUILD", `Expected BUILD domain, received ${event.domain}`);

  const decoded = decode(event);
  console.assert(decoded.meaning === "BUILD request", `Unexpected meaning: ${decoded.meaning}`);
  console.assert(decoded.text === "i m goin 2 bld da sys 2moro coz i nd 2 knw hw it works");

  const log = encodeLog([
    { text: "please build the model", role: "user" },
    { text: "I can help with that", role: "assistant" }
  ]);
  console.assert(log.length === 2, "encodeLog should encode all messages");

  const context = logToContext(log);
  console.assert(context.includes("[U ch:0"), "Context should serialise user events");

  const pianoRoll = toPianoRoll(log, 4);
  console.assert(pianoRoll.includes("*"), "Piano roll should include event markers");

  console.log("midi-encoder.js tests passed");
}

File: mobile/lib/midi_encoder.dart
import 'app_models.dart';
import 'midi_grammar.dart';
import 'mxit_compressor.dart';

class DecodedEvent {
  const DecodedEvent({
    required this.role,
    required this.meaning,
    required this.intensity,
    required this.importance,
    required this.text,
  });

  final String role;
  final String meaning;
  final String intensity;
  final String importance;
  final String text;
}

final Set<int> _actionDomains = {
  pitchDomain['REQUEST']!,
  pitchDomain['COMMAND']!,
  pitchDomain['BUILD']!,
  pitchDomain['FIX']!,
  pitchDomain['SEARCH']!,
  pitchDomain['SEND']!,
  pitchDomain['SAVE']!,
};

String _normaliseRole(MidiEvent event) {
  if (event.role.isNotEmpty) {
    return event.role.toLowerCase();
  }

  return (channelName[event.channel] ?? 'USER').toLowerCase();
}

String _describeMeaning(MidiEvent event, String role) {
  final domain = event.domain;

  if (role == 'user' && _actionDomains.contains(event.pitch)) {
    return '$domain request';
  }
  if (role == 'assistant') {
    return '$domain response';
  }
  if (event.pitch == pitchDomain['QUESTION']) {
    return 'QUESTION prompt';
  }

  return '$domain message';
}

String _describeIntensity(int velocity) {
  if (velocity >= velocityLevels['URGENT']!) {
    return 'URGENT ($velocity)';
  }
  if (velocity >= velocityLevels['EMPHASIS']!) {
    return 'EMPHASIS ($velocity)';
  }
  if (velocity >= velocityLevels['NORMAL']!) {
    return 'NORMAL ($velocity)';
  }
  if (velocity >= velocityLevels['SOFT']!) {
    return 'SOFT ($velocity)';
  }
  if (velocity >= velocityLevels['WHISPER']!) {
    return 'WHISPER ($velocity)';
  }
  return 'SILENT ($velocity)';
}

String _describeImportance(int duration) {
  if (duration >= durationLevels['ANCHOR']!) {
    return '${durationLevels['ANCHOR']}ms';
  }
  return '$duration' 'ms';
}

MidiEvent encode(String rawText, String role) {
  final compressed = mxitCompress(rawText);
  return buildEvent(rawText, compressed.text, role);
}

DecodedEvent decode(MidiEvent event) {
  final role = _normaliseRole(event);
  return DecodedEvent(
    role: role,
    meaning: _describeMeaning(event, role),
    intensity: _describeIntensity(event.velocity),
    importance: _describeImportance(event.duration),
    text: event.compressed.isNotEmpty ? event.compressed : mxitCompress(event.raw).text,
  );
}

List<MidiEvent> encodeLog(List<Map<String, String>> messages) {
  return messages.map((message) {
    return encode(
      message['text'] ?? '',
      message['role'] ?? 'user',
    );
  }).toList(growable: false);
}

String logToContext(List<MidiEvent> events) {
  return events.map(serialiseEvent).join('\n');
}

String toPianoRoll(List<MidiEvent> events, {int width = 32}) {
  if (events.isEmpty) {
    return '(empty piano roll)';
  }

  final columnCount = width < 1 ? 1 : width;
  final visibleEvents = events.length > columnCount
      ? events.sublist(events.length - columnCount)
      : [...events];
  final pitches = visibleEvents.map((event) => event.pitch).toSet().toList()
    ..sort((left, right) => right.compareTo(left));

  return pitches.map((pitch) {
    final label = pitchToNoteName(pitch).padRight(3);
    final columns = visibleEvents
        .map((event) => event.pitch == pitch ? '*' : ' ')
        .join()
        .padRight(columnCount);
    return '$label | $columns |';
  }).join('\n');
}
